### ISSUE TITLE
Question: Compilation of Server Bundle

### ISSUE THREAD

**Issue by dtothefp** [halt-hammerzeit/webpack-isomorphic-tools#7]
This is probably a stupid question but I'm just trying to figure out how to get up and running with this module as it looks very useful, but I'm confused at what the `server` method does. When I pass my `./server` entry point as the second argument to `new webpack-isomorphic-tools()`, what is actually happening? Is it just relying on a webpack compilation of client code to have happened that output a webpack `stat.json` file into the root of the `context` directory, or does `webpack-isomorphic-tools` actually bundle (with webpack) and output my `./server` bundle in a `target=node` webpack config environment. Sorry for my confusion but not sure I'm understanding the functionality of this module completely.

**Comment by halt-hammerzeit** [comment]
Well, to begin with, `new webpack-isomorphic-tools()` no more has a second argument (for a long time now).
When you call the `server` method it waits for stats json to be generated (for the first time), then activates `require()` hooks for the extensions you specified, and then runs your `./server` entry point.
It just relies on that stats json file, and that's all.
It even doesn't know a thing called `Webpack` exists and it doesn't cooperate with Webpack or Webpack generated code.

Its counterpart though does because it's a webpack plugin, and its purpose is to generate that stats json file.

I didn't know about `target node` you pointed out.
That's an interesting reading.
http://jlongster.com/Backend-Apps-with-Webpack--Part-I

**Comment by dtothefp** [comment]
@halt-hammerzeit thanks for the info. Just to clarify, the second argument I was talking about was the `./server` entry point path passed to the the `server` method of the `webpack-isomorphic-tools` instance.

I have read that article you referenced and use that technique along with the demonstration by @sokra for generating a `node` side server bundle https://github.com/webpack/react-webpack-server-side-example/blob/master/webpack.config.js#L38.

The reason why I was confused with your plugin/tools is that I see that your [webapp](https://github.com/halt-hammerzeit/webapp/blob/master/code/server/entry.js) example and the [react-redux-universal-hot-example](https://github.com/erikras/react-redux-universal-hot-example/blob/master/bin/server.js)  seem to both write the server side bundle entry point in es5, but then the actual `server` code is written in es6. I guess you guys are just compiling that bundle with babel to run it, and I was trying to pre-compile with webpack so my css/scss requires weren't breaking. Looks like that's what `webpack-isomorphic-tools` looks to solve by accessing the stats from the `compiler('done'` event. 

Anyhow, still trying to wrap my head around all of this and figure out the best way to move forward with my project. Am I correct that if I were using the `webpack-isomorphic-tools` I would not directly require/import css/scss into the file but would rather pass the references css/scss stats down from the top level component. If so it would probably be nice to write some decorators similar to https://github.com/jordangarcia/nuclear-js-react-addons that could pass and retrieve asset data from react context.

**Comment by halt-hammerzeit** [comment]
`webapp` and the universal example are written in ES6.
But node.js doesn't support ES6 (afaik) so Babel is used (not to mention the benefit of a ton of plugins, for example the JSX plugin).
All client code is compiled with babel during webpack build when `babel-loader` activates for `.js` files.
On the server side we don't use Webpack (i have no need for that) so we activate Babel in those `entry` .js files (their sole purpose is to run `babel/register`).
So yes, we are compiling it with Babel, in run time (done once for each module).

`webpack-isomorphic-tools` was introduced to make `require()` of *.css (for example) files not throw an exception on the server (and furthermore, for it to return something meaningful like the path or the complied stylesheet).
`stats` are used for that purpose, you got that one right.
But since I didn't know of the possibility of `target node` I didn't test this module with it.
I can do that if you provide me a simple example project which I can test upon (with the instructions).

I didn't understand what you're saying in the last paragraph.
Maybe I could if you clarify what you mean there.

**Comment by dtothefp** [comment]
Yeh sorry, last paragraph is stupid. I see that you are getting stats for extracted css bundle names and able to template them in as `<link>` tags in JSX. One thing I'm confused about is if Babel bundles your entire component, and in your README example for this repo, you pass the `component` as a `prop`

```jsx
<div id="content" dangerouslySetInnerHTML={{__html: React.renderToString(component)}}/>
```
babel will be compiling that component chain as well and will it not throw an error if it sees you require/import css/scss in any components in that tree? I saw in your README you say you don't have to use the css-loader `module` feature for local scoping of css but wasn't sure how you were accomplishing it through the plugin? Also, saw in the redux example they were using less/scss but didn't quite see how it was working. Some clarification around this would be great

**Comment by halt-hammerzeit** [comment]
babel will be compiling that component chain and will not throw `SyntaxError`s or whatever because the compilation happens at runtime in this case and in runtime we have the `require()` hooks activated for all the assets (css and such).

**Comment by halt-hammerzeit** [comment]
I can see how compile time `require()` hooks could possibly work too: using that `NormalModuleReplacementPlugin` in case it runs before Babel transpiles a file.
It's a hypothesis.

If you have any questions left I can clarify them.

**Comment by dtothefp** [comment]
Sorry for the delay, I'm on eastern timezone. Thanks for the explanation of babel/register require hooks. I wasn't aware completely of how it worked. Maybe my usecase is different and that is why I'm not understanding completely.
- I would like to compile my react component tree to HTML and extract into a separate file because I will be serving this content statically. Is it still possible or useful to use `isomorpic-tools`?
- It seems your examples and those in redux are for serving HTML content from express application server so I'm not sure how/if this relates to my project.

Essentially, I'm still confused because of the example in redux

https://github.com/erikras/react-redux-universal-hot-example#styles
https://github.com/erikras/react-redux-universal-hot-example/blob/0af04acb9a0072ac1581e23653327d0e5011cd0e/src/components/InfoBar/InfoBar.js

I'm assuming that is why the `require` is used inside of the `render` method, is this what you were saying about "compilation happens at runtime"? Would this then fail if I tried to `import styles from './whatevs.scss'`?

**Comment by halt-hammerzeit** [comment]
1. Yes, of course, that's what it's ment to do: to render React components on the server-side.
2. Aren't you using `express`? If you just use `fs.writeFileSync`, then it doesn't matter: you just move your project to `express` and the use Linux shell with `wget` to download the markup (or you can write a small Node.js application to dowload contents of a URL)

Regarding `import` vs `require` there's a section in the README, go read it.
In short: `import` does the job but `require` is for hot reload to work (while you're developing your project); `import` is equivalent to `require` for production.

**Comment by dtothefp** [comment]
yeh I read that about the import vs require. I think I was confused about `require` hooks. Just saw this method `register_extension(extension)` so I am understanding more. I won't be using express except for a dev server. I'm essentially using a static template generator to generate the layout markup in another framework other than react `nunjucks` and then will include the compiled react component into that as a partial. I'll put together a small demo repo, but I think I understand more of how to do this with `isomorphic-tools` now

**Comment by halt-hammerzeit** [comment]
@dtothefp ok.
If you have any suggestions about rewriting the README you can post them to me because I may have a blind spot for documentation issues being the project creator.

If you won't be using `express` then you don't need it. You can just take the code from the `render()` function and adapt it to your needs.
https://github.com/halt-hammerzeit/webapp/blob/master/code/server/webpage%20rendering.js
https://github.com/halt-hammerzeit/webapp/blob/master/code/server/web%20(koa).js#L80-L86

**Comment by halt-hammerzeit** [comment]
Or better look through just these two files
https://github.com/halt-hammerzeit/webapp/tree/master/code/react-isomorphic-render

There will be lines you won't understand but overall you can get the idea from these two files.
If you'll have questions on the code you can post them here.

**Comment by dtothefp** [comment]
@halt-hammerzeit thanks for all the explanations, I think I finally got it and tried to mix what I saw in the webapp example with the redux example. I made a demo here https://github.com/dtothefp/isomorphic-demo/tree/dfp/add-assemble

with the basic express server middleware here https://github.com/dtothefp/isomorphic-demo/blob/dfp/add-assemble/app/index.js

Essentially, I'm generating html using `nunjucks` and a static template generator `assemble`. Then I template the html for the react component and inject it in the template. The idea of this is that in local dev it will work via express middleware, and in production build it will most likely be run by something like an AWS lamba, and the html will be written to disk and uploaded. 

Anyhow, I still have a few questions, not all related to this plugin but general webpack build stuff as well so as I noticed a few new things that I wasn't currently using:
- I'm getting an error in `NODE_ENV="development"` mode related to local css that is not extracted. I know you said local styles won't work when not extracted (and from what I've seen extraction doesn't occur using the `webpack-dev-middleware`, but I don't see these errors in your webpapp repo or the redux repo so I'm wondering how to eliminate them?
terminal:
<img width="819" alt="screen shot 2015-10-20 at 1 00 16 am" 
src="https://cloud.githubusercontent.com/assets/4656726/10598333/f64b8fba-76c5-11e5-8cd3-cbd1566ca133.png">

console: 
<img width="615" alt="screen shot 2015-10-20 at 1 00 52 am" src="https://cloud.githubusercontent.com/assets/4656726/10598339/0b5b2140-76c6-11e5-9b1d-3f70ca86dbf2.png">
- also for the reason above when building a bundle of SASS that will be used globally, I build it in a separate webpack build that does not use the dev-middleware and instead uses `compiler.watch` (because I don't want to load all the css into js because it causes "flickering" locally). I'm wondering if you have some tricks around this?
- also I started using the `'react-transform'` plugin, as I'm assuming it is a replacement somehow for the `react-hot-loader`. I used the following config but if I include `rewire` plugin I get an error inside `babel-rewire` code. 
```
Uncaught ReferenceError: _ReactTemp is not defined
```

I can put an issue on `'react-transform'` or `rewire` but wondering if you know a solution off hand?

```js
    const babelDevConfig = {
      plugins: [
        'rewire',
        'react-transform'
      ],
      extra: {
        'react-transform': {
          transforms:
            [{
              transform: 'react-transform-hmr',
              imports: ['react'],
              locals: ['module']
            }]
        }
      }
    };
```

**Comment by halt-hammerzeit** [comment]
Okay, I see, you figured the things out.
As for the webpack-isomophic-tools error, the fix will be:

```
    style_modules: {
      extensions: ['css', 'less','scss'],
      filter: function(m, regex, options, log) {
        if (!options.development) {
          return m.name.indexOf('./~/css-loader') === 0 && regex.test(m.name);
        }
        //filter by modules with '.scss' inside name string, that also have name and moduleName that end with 'ss'(allows for css, less, sass, and scss extensions)
        //this ensures that the proper scss module is returned, so that namePrefix variable is no longer needed
        return m.name.indexOf('./~/css-loader') === 0 && (regex.test(m.name) && m.name.slice(-2) === 'ss' && m.reasons[0].moduleName.slice(-2) === 'ss');
      },
```

As you can see, I added a `m.name.indexOf('./~/css-loader') === 0`
This way the style object isn't overwritten with an empty object.

After fixing this error I'm still unable to browse your application because it says
`TypeError: Cannot read property 'render' of undefined` here:

```
      const page = assemble.pages.getView('pages/index');
      page.render({
```

for some reason `page` is undefined.
I don't have any experience with that `assemble` module so you should figure out why that `.getView()` returns undefined.

If you'll any issues after fixing that one I can help too.

**Comment by dtothefp** [comment]
@halt-hammerzeit thanks for the feedback. I just updated the master branch of my repo with your fix so you can reference here

https://github.com/dtothefp/isomorphic-demo

I only had to add your `indexOf` check for `development` otherwise my app would break as the css styles in the `webpack-stats.json` wouldn't be generated.

Also, for the assemble problem could you try `npm i -D assemble-loader@latest` and let me know if it works? Not sure what the issue is, but I'm working off their dev branch so anything is possible :neckbeard:. @doowb could you have a look?

**Comment by halt-hammerzeit** [comment]
Oh, yeah, that's right, I didn't think of checking that quickfix in production mode.

I installed the latest version but it doesn't work.
```
^CG:\work\isomorphic-demo [master +1 ~0 -0 !]> npm i -D assemble-loader@latest
npm WARN install Couldn't install optional dependency: Unsupported
assemble-demo@0.1.0 G:\work\isomorphic-demo
└── assemble-loader@0.2.2

npm WARN EPACKAGEJSON assemble-demo@0.1.0 license should be a valid SPDX license expression
```

Still the same exception.

I've looked through your code a bit and there seems to be a possibility of file separator issue.

```
  assemble.create('pages', {renameKey: function(fp) {
    const dirname = path.dirname(fp).split('/').slice(-1)[0];
    const basename = path.basename(fp).split('.').slice(0)[0];

    return `${dirname}/${basename}`;
  }})
  .use(loader());
```

To begin with, i'm on Windows, so it's not `/`, it's `\`.
Then, the assemble library itself may be relying on `/` instead of `\` and end up not finding things by such incorrect paths»

**Comment by dtothefp** [comment]
hmmmm @doowb is an ex .NET guy so I'll check with him.

**Comment by dtothefp** [comment]
PS if I re-order the `rewire` plugin to go after the `react-transform` then everything works. Just an FYI, I'll put an issue on the `react-tranform-hmr` repo

```js
    const babelDevConfig = {
      plugins: [
        'react-transform',
        'rewire'
      ],
      extra: {
        'react-transform': {
          transforms:
            [{
              transform: 'react-transform-hmr',
              imports: ['react'],
              locals: ['module']
            }]
        }
      }
    };
```

**Comment by doowb** [comment]
assemble works with both windows and unix style path separators. @dtothefp if that's your code in the renameKey function, then that just needs to be updated to use the correct separator for the environment.

**Comment by dtothefp** [comment]
@doowb @halt-hammerzeit updated the `renameKey` function if that was the culprit

https://github.com/dtothefp/isomorphic-demo/commit/9b536cfead4be8d8d0133f7c0f1003eef33e46df#diff-2

**Comment by halt-hammerzeit** [comment]
@dtothefp I pulled your changes, done `npm install` and `gulp watch` - the error is the same

**Comment by doowb** [comment]
If you're referring to this error:

> TypeError: Cannot read property 'render' of undefined

Try doing `console.log(assemble.pages.views)` just before the `const page = assemble.pages.getView()` line. This will tell you if any pages were actually loaded. The globby might not be finding the pages correct, so none are getting loaded onto the views.

**Comment by halt-hammerzeit** [comment]
@doowb Here it is

```
{ 'G:\work\isomorphic-demo\src\templates\pages\index': <File "index.html" <Buffer 2d 2d 2d 0d 0a 74 69 74 6c 65 3a 20 4d
 61 69 6e 20 50 61 67 65 20 54 69 74 6c 65 0d 0a 2d 2d 2d 0d 0a 0d 0a 7b 25 20 65 78 74 65 6e 64 73 20 6c 61 79 ... >> }
```

**Comment by doowb** [comment]
Thanks @halt-hammerzeit . @dtothefp I commented on the commit for the `renameKey`. I think you just need to use a regex to split instead of `'/'`. Then you'll get the correct segments to build the key.

**Comment by dtothefp** [comment]
alright @doowb and @halt-hammerzeit sorry for the confusion, I have never done `node` work on windows.

updated to use `path.sep`

```js
export default function(fp) {
  const dirname = path.dirname(fp).split(path.sep).slice(-1)[0];
  const basename = path.basename(fp, path.extname(fp));
  return path.join(dirname, basename);
}
```

**Comment by halt-hammerzeit** [comment]
@dtothefp ok, I can run it now and it renders.
did you have any questions regarding rendering?

**Comment by dtothefp** [comment]
@halt-hammerzeit nope, this is good for my use case. Thanks for answering all of my extra questions that were off topic of this issue.

**Comment by halt-hammerzeit** [comment]
FYI
I'm posting this in every issue and PR to notify whoever may be interested:
today I've released an alternative helper library called [`universal-webpack`](https://github.com/halt-hammerzeit/universal-webpack).
It takes a different approach than `webpack-ismorphic-tools` and instead of hacking Node.js `require()` calls it just compiles all code with `target: 'node'` webpack configuration option.
As a result, all Webpack plugins and features are supported.
If you think you might need that here's an example project:
https://github.com/halt-hammerzeit/webpack-react-redux-isomorphic-render-example
